File: algorithm/lc234_palindrome_linked_list.py
"""
    https://leetcode.com/problems/palindrome-linked-list/
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True

        fast, slow = head

        while fast.next and fast.next.next:
            slow, fast = slow.next, fast.next.next

        """
            start reversing from the `slow.next` pointer
            get the new `head` pointer of the reversed list
            then assign the new `head` pointer to `slow`
        """
        slow = self.reverseList(slow.next)

        """
            compare the head and slow head list
        """
        while slow:
            if slow.val != head.val:
                return False
            slow = slow.next
            head = head.next

        return True

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur, prev = None
        while cur:
            cur.next, prev, cur = prev, cur, head.next

            # next = head.next
            # head.next = prev
            # prev = head
            # head = next
        return prev

File: algorithm/lc234_palindrome_linked_list.go
/**
    https://leetcode.com/problems/palindrome-linked-list/
**/
package algorithm

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	fast, slow := head, head

	for fast.Next != nil && fast.Next.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
	}

	slow = reverseList(slow.Next)

	for slow != nil {
		if slow.Val != head.Val {
			return false
		}
		slow, head = slow.Next, head.Next
	}

	return true

}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode

	for head != nil {
		// next := head.Next
		// head.Next = prev
		// prev = head
		// head = next

		head.Next, prev, head = prev, head, head.Next
	}
	return prev
}
